src/circular_buffer.rs

```diff
@@ -0,0 +1,313 @@
+
+#![deny(missing_docs)]
+#![deny(warnings)]
+
+use core::marker::PhantomData;
+
+/// A circular buffer
+pub struct CircularBuffer<T, A>
+where
+    A: AsMut<[T]> + AsRef<[T]>,
+    T: Copy,
+{
+    _marker: PhantomData<[T]>,
+    array: A,
+    index: usize,
+    len: usize,
+}
+
+impl<T, A> CircularBuffer<T, A>
+where
+    A: AsMut<[T]> + AsRef<[T]>,
+    T: Copy,
+{
+    /// Creates a new empty circular buffer using `array` as backup storage.
+    ///
+    /// # Example
+    /// ```
+    /// #![feature(const_fn)]
+    /// use heapless::CircularBuffer;
+    /// const LEN: usize = 16;
+    /// // with let
+    /// let buf_let = CircularBuffer::new([0;LEN]);
+    /// // with const
+    /// const BUF_CONST: CircularBuffer<u8, [u8; LEN]>
+    ///     = CircularBuffer::new([0;LEN]);
+    /// // with static
+    /// static BUF_STATIC: CircularBuffer<u8, [u8; LEN]>
+    ///     = CircularBuffer::new([0;LEN]);
+    /// // with slice
+    /// let mut array = [0_usize; 8];
+    /// let mut buf_slice = CircularBuffer::new(&mut array);
+    /// ```
+    #[inline]
+    pub const fn new(array: A) -> Self {
+        CircularBuffer {
+            _marker: PhantomData,
+            array: array,
+            index: 0,
+            len: 0,
+        }
+    }
+
+    /// Creates a new circular buffer using `array` as backup storage.
+    /// Keeps the values and the length of the array.
+    /// The backup storage is completly used upon creation.
+    /// Insertions will overwrite at the start of the array.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let buf = CircularBuffer::from_array([1,2,3]);
+    /// assert!(buf.is_full());
+    /// ```
+    #[inline]
+    pub fn from_array(array: A) -> Self {
+        let len = array.as_ref().len();
+        CircularBuffer {
+            _marker: PhantomData,
+            array: array,
+            index: 0,
+            len: len,
+        }
+    }
+
+    /// Returns the capacity of this buffer.
+    ///
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// const LEN: usize = 16;
+    /// let buf = CircularBuffer::new([0;LEN]);
+    /// assert_eq!(buf.capacity(), LEN);
+    /// ```
+    #[inline]
+    pub fn capacity(&self) -> usize {
+        self.array.as_ref().len()
+    }
+
+    /// Returns the current length of this buffer.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let buf = CircularBuffer::from_array([1,2,3]);
+    /// assert_eq!(buf.len(), 3);
+    /// ```
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns the current index of this buffer.
+    /// This is the next place to push to.
+    ///
+    /// # Example
+    /// ```
+    /// use heapless::CircularBuffer;
+    /// let buf = CircularBuffer::from_array([1,2,3]);
+    /// assert_eq!(buf.index(), 0);
+    /// ```
+    #[inline]
+    pub fn index(&self) -> usize {
```

Comment:
I'm not sure if we want to expose this and the `first_index` method in the API. Do you have an use case for it?

Reply:
It is used in the circular_buffer::Iter::new method to initialize the iterator.
It could be made non-public.